File: problems/123.py
class Solution:
    def maxProfit(self, prices):
        buy = [0] * 2
        sell = [0] * 2
        for i in range(len(prices)):
            if i == 0:
                buy[0] = 0 - prices[i]
            elif i == 1:
                buy[0] = max(buy[0], 0 - prices[i])
                sell[0] = max(sell[0], prices[i] + buy[0])
            else:
                buy[0] = max(buy[0], 0 - prices[i])
                sell[0] = max(sell[0], prices[i] + buy[0])

                buy[1] = max(buy[1], sell[0] - prices[i])
                sell[1] = max(sell[1], prices[i] + buy[1])

        return sell[1]

File: problems/dp/123.go
package dp

import "math"

func maxProfit(prices []int) int {
	buy1, buy2 := math.MinInt64, math.MinInt64
	sell1, sell2 := 0, 0
	for _, v := range prices {
		if buy1 < 0 - v {
			buy1 = 0 - v
		}
		if sell1 < buy1 + v {
			sell1 = buy1 + v
		}
		if buy2 < sell1 - v {
			buy2 = sell1 - v
		}
		if sell2 < buy2 + v {
			sell2 = buy2 + v
		}
	}
	return sell2
}
